internal/scan/scan_partition.go

```diff
@@ -0,0 +1,87 @@
+package scan
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/stellar/go/support/datastore"
+)
+
+// scanPartition walks a partition in DESC order and records gaps.
+func scanPartition(
+	ctx context.Context,
+	partition Partition,
+	ds datastore.DataStore,
+	schema datastore.DataStoreSchema,
+) (Result, error) {
+	res := Result{gaps: make([]Gap, 0)}
+
+	// StartAfter normally uses (high+1); if high==MaxUint32, leave empty to mean "no upper bound".
+	startAfter := ""
+	if partition.high != math.MaxUint32 {
+		startAfter = schema.GetObjectKeyFromSequenceNumber(partition.high + 1)
+	}
+
+	it := &LedgerFileIter{
+		DS:         ds,
+		StartAfter: startAfter, // empty => start from the highest possible key
+		StopAfter:  schema.GetObjectKeyFromSequenceNumber(partition.low),
+	}
+
+	for cur, err := range it.Next(ctx) {
+		if err != nil {
+			return res, err
+		}
+		if err := ctx.Err(); err != nil {
+			return res, err
+		}
```

Comment:
no need to check the context error here because any operations that are potentially long lived take a context and will return immediately once the context is done